sklearn/utils/_unique.py

```diff
@@ -0,0 +1,96 @@
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+import numpy as np
+
+from sklearn.utils._array_api import get_namespace
+
+
+def _attach_unique(y):
+    """Attach unique values of y to y and return the result.
+
+    The result is a view of y, and the metadata (unique) is not attached to y.
+    """
+    if not isinstance(y, np.ndarray):
+        return y
+    try:
+        # avoid recalculating unique in nested calls.
+        if "unique" in y.dtype.metadata:
+            return y
+    except (AttributeError, TypeError):
+        pass
+
+    unique = np.unique(y)
+    unique_dtype = np.dtype(y.dtype, metadata={"unique": unique})
+    return y.view(dtype=unique_dtype)
+
+
+def attach_unique(*ys, return_tuple=False):
+    """Attach unique values of ys to ys and return the results.
+
+    The result is a view of y, and the metadata (unique) is not attached to y.
```

Comment:
Can we include a comment here that starts that the output of `attach_unique` should never be returned from a public function?

Reply:
Yep, added a comment for this.